### ISSUE TITLE
DynamicRealmObject.setList() does wrong type check

### ISSUE THREAD

**Issue by jonasrottmann** [realm/realm-java#4252]
#### Goal

Set the value of a field with the type RealmList via `DynamicRealmObject.setList()`.

#### Expected Results

The value is set.

#### Actual Results

`java.lang.IllegalArgumentException: The elements in the list is not the proper type. Was EmailAddress expected class_User.`
The type of the items of the passed in list are checked against the type of the table. 

#### Code Sample
```java
public class User extends RealmObject {
    RealmList<EmailAddress> emailAddresses;

    public User() {

    }
}
```
```java
public class EmailAddress extends RealmObject {
    @PrimaryKey
    String emailAddress;

    public EmailAddress() {

    }
}
```
```java
dynamicRealm.beginTransaction();

DynamicRealmObject user = dynamicRealm.createObject(User.class.getSimpleName());
DynamicRealmObject email = dynamicRealm.createObject(EmailAddress.class.getSimpleName(), "email");
user.getList("emailAddresses").add(email);

DynamicRealmObject user2 = dynamicRealm.createObject(User.class.getSimpleName());
user2.setList("emailAddresses", user.getList("emailAddresses"));

dynamicRealm.commitTransaction();
```

#### Version of Realm and tooling
Realm version(s): 2.3.1

Realm sync feature enabled: no

Android Studio version: 2.2.3

Which Android version and device: Android 7.1.2 Nexus 5X

**Comment by Zhuinden** [comment]
``` java
dynamicRealm.beginTransaction();

DynamicRealmObject user = dynamicRealm.createObject(User.class.getSimpleName());
DynamicRealmObject email = dynamicRealm.createObject(EmailAddress.class.getSimpleName(), "email");
user.getList("emailAddresses").add(email);

DynamicRealmObject user2 = dynamicRealm.createObject(User.class.getSimpleName());
RealmList<DynamicRealmObject> emails = user2.getList("emailAddresses");
for(DynamicRealmObject email : user.getList("emailAddresses")) {
    emails.add(email);
}
dynamicRealm.commitTransaction();
```

**Comment by zaki50** [comment]
Hi @jonasrottmann 

There seems to be a type checking bug in `DynamicRealmObject#setList()`.
I'll check the code.

Current work-around is using `addAll()` as @Zhuinden said.

**Comment by zaki50** [comment]
fixed by #4254
